### PR TITLE
feat: Updates singular `mongodbatlas_advanced_cluster` data source to support independent shard scaling & updates relevant flattener methods

### PR DESCRIPTION
**Pull request by maastha**
## Description

This PR updates singular `mongodbatlas_advanced_cluster` data source to support new independent shard scaling schema & updates relevant flattener methods to use the latest SDK objects wherever possible.
To continue supporting current API objects, I have added conversion methods in `model_sdk_version_conversion.go`

Link to any related issue(s): [CLOUDP-257179](https://jira.mongodb.org/browse/CLOUDP-257179)

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [ ] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [ ] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [ ] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [ ] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [ ] I have added any necessary documentation (if appropriate)
- [ ] I have run make fmt and formatted my code
- [ ] If changes include deprecations or removals, I defined an isolated PR with a relevant title as it will be used in the auto-generated changelog.
- [ ] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

File: internal/service/advancedcluster/model_advanced_cluster.go [hunk 1]
```diff
@@ -933,3 +951,49 @@ func expandRegionConfigAutoScaling(tfList []any) *admin.AdvancedAutoScalingSetti
 	}
 	return &settings
 }
+
+func flattenAdvancedReplicationSpecsDS(ctx context.Context, apiRepSpecs []admin.ReplicationSpec20240710, d *schema.ResourceData, connV2 *admin.APIClient) ([]map[string]any, error) {
```
Comment (maastha) [inline internal/service/advancedcluster/model_advanced_cluster.go:956]: I've split the flattener for replication spec for data source separately for now to keep the logic simple. Any improvements/refactoring would be done once update & relevant resource changes are implemented

File: internal/service/advancedcluster/data_source_advanced_cluster.go [hunk 2]
```diff
@@ -235,50 +250,114 @@ func DataSource() *schema.Resource {
...
 		}
-		return diag.FromErr(fmt.Errorf(errorRead, clusterName, err))
 	}
 
+	if err := d.Set("replication_specs", replicationSpecs); err != nil {
+		return diag.FromErr(fmt.Errorf(ErrorClusterAdvancedSetting, "replication_specs", clusterName, err))
+	}
+
+	// TODO: update to use connLatest to call below API
+	processArgs, _, err := connV2.ClustersApi.GetClusterAdvancedConfiguration(ctx, projectID, clusterName).Execute()
```
Comment (maastha) [inline internal/service/advancedcluster/data_source_advanced_cluster.go:322]: Looks like return type of `GetClusterAdvancedConfiguration` call has changed from `*admin.ClusterDescriptionProcessArgs` to `map[string]interface{}` in the latest SDK

File: internal/service/advancedcluster/resource_advanced_cluster.go [hunk 3]
```diff
@@ -339,6 +341,11 @@ func schemaSpecs() *schema.Schema {
 		Optional: true,
 		Elem: &schema.Resource{
 			Schema: map[string]*schema.Schema{
+				"disk_size_gb": { // new API only
```
Comment (maastha) [inline internal/service/advancedcluster/resource_advanced_cluster.go]: common schema method between data source & resource

File: internal/service/advancedcluster/data_source_advanced_cluster.go (hunk 2 above)
Comment (maastha) [reply internal/service/advancedcluster/data_source_advanced_cluster.go:322]: will look more into this as part of https://jira.mongodb.org/browse/CLOUDP-258709, will continue using the old SDK for this API call for now

**Comment by maastha** [comment]
Acceptance tests will be updated in a separate PR

File: internal/service/advancedcluster/model_advanced_cluster.go [hunk 4]
```diff
@@ -497,24 +511,28 @@ func doesAdvancedReplicationSpecMatchAPI(tfObject map[string]any, apiObject *adm
...
-	d *schema.ResourceData, connV2 *admin20231115.APIClient) (map[string]any, error) {
+func flattenAdvancedReplicationSpecOldSDK(ctx context.Context, apiObject *admin20231115.ReplicationSpec, tfMapObject map[string]any,
+	d *schema.ResourceData, connV2 *admin.APIClient) (map[string]any, error) {
 	if apiObject == nil {
 		return nil, nil
 	}
+	var rootDiskSizeGB *float64
+	if v, ok := d.GetOk("disk_size_gb"); ok {
+		rootDiskSizeGB = conversion.Pointer(v.(float64))
+	}
```
Comment (AgustinBettati) [inline internal/service/advancedcluster/model_advanced_cluster.go]: I believe this was copied from the expander but does not work in the context of a flattener, we would have to pass the value define in the root of the api response

File: internal/service/advancedcluster/model_advanced_cluster.go [hunk 5]
```diff
@@ -644,6 +663,9 @@ func flattenAdvancedReplicationSpecRegionConfigSpec(apiObject *admin20231115.Ded
 				tfMap["ebs_volume_type"] = apiObject.GetEbsVolumeType()
 			}
 		}
+		if _, ok := tfMapObject["disk_size_gb"]; ok {
+			tfMap["disk_size_gb"] = apiObject.GetDiskSizeGB()
+		}
```
Comment (AgustinBettati) [inline internal/service/advancedcluster/model_advanced_cluster.go:664]: do we need to check if the attribute is defined in the config? Since it is compute I see not issue in storing the result in the state always

File: internal/service/advancedcluster/data_source_advanced_cluster.go [hunk 6]
```diff
@@ -235,50 +250,119 @@ func DataSource() *schema.Resource {
...
+			}
+			return diag.FromErr(fmt.Errorf(errorRead, clusterName, err))
+		}
+
+		clusterID = clusterDescLatest.GetId()
+
+		replicationSpecs, err = flattenAdvancedReplicationSpecsDS(ctx, clusterDescLatest.GetReplicationSpecs(), d, connLatest)
+		if err != nil {
+			return diag.FromErr(fmt.Errorf(ErrorClusterAdvancedSetting, "replication_specs", clusterName, err))
+		}
```
Comment (AgustinBettati) [inline internal/service/advancedcluster/data_source_advanced_cluster.go:309]: within this path, would we need to also disk_size_gb by setting the value defined in the nested hardware spec object?

File: internal/service/advancedcluster/model_advanced_cluster.go (hunk 4 above)
Comment (maastha) [reply internal/service/advancedcluster/model_advanced_cluster.go]: this does work because I set it in the schema beforehand. But I see that could be hard to follow, let me refactor

File: internal/service/advancedcluster/model_advanced_cluster.go (hunk 5 above)
Comment (maastha) [reply internal/service/advancedcluster/model_advanced_cluster.go:664]: I have maintained the previous logic here so we should be able to use this for the resource as well

Main logic change for flatteners is the removal of replication spec matching logic for the data source as we should handle that separately for the resource since now we also need to accommodate two APIs. Everything else is similar to current logic

File: internal/service/advancedcluster/data_source_advanced_cluster.go (hunk 6 above)
Comment (maastha) [reply internal/service/advancedcluster/data_source_advanced_cluster.go:309]: you mean set the root level disk_size_gb based on the nested value?

File: internal/service/advancedcluster/model_advanced_cluster.go (hunk 4 above)
Comment (AgustinBettati) [reply internal/service/advancedcluster/model_advanced_cluster.go]: ohh I see, had missed that

File: internal/service/advancedcluster/data_source_advanced_cluster.go (hunk 6 above)
Comment (AgustinBettati) [reply internal/service/advancedcluster/data_source_advanced_cluster.go:309]: yes was thinking this case

File: internal/service/advancedcluster/model_advanced_cluster.go (hunk 4 above)
Comment (maastha) [reply internal/service/advancedcluster/model_advanced_cluster.go]: have refactored

File: internal/service/advancedcluster/data_source_advanced_cluster.go (hunk 6 above)
Comment (maastha) [reply internal/service/advancedcluster/data_source_advanced_cluster.go:309]: discussed offline, will keep as is for now

**Review by AgustinBettati — APPROVED** [review]
LGTM
